src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/CardinalityBasedGreedyJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,291 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.poptimizer.simple;
+
+import se.liu.ida.hefquin.engine.federation.BRTPFServer;
+import se.liu.ida.hefquin.engine.federation.FederationMember;
+import se.liu.ida.hefquin.engine.federation.SPARQLEndpoint;
+import se.liu.ida.hefquin.engine.federation.TPFServer;
+import se.liu.ida.hefquin.engine.federation.access.*;
+import se.liu.ida.hefquin.engine.federation.access.impl.req.BRTPFRequestImpl;
+import se.liu.ida.hefquin.engine.federation.access.impl.req.TPFRequestImpl;
+import se.liu.ida.hefquin.engine.federation.access.utils.FederationAccessUtils;
+import se.liu.ida.hefquin.engine.federation.access.utils.RequestMemberPair;
+import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalOperator;
+import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.physical.impl.*;
+import se.liu.ida.hefquin.engine.queryplan.utils.PhysicalPlanFactory;
+import se.liu.ida.hefquin.engine.queryproc.PhysicalOptimizationException;
+
+import java.util.*;
+
+/**
+ * This class implements a query optimizer that builds left-deep query plans,
+ * for which it uses a greedy approach to determine the join order based on cardinality estimation,
+ * and then choose physical algorithm according to the estimated number of request to execute the join.
+ */
+public class CardinalityBasedGreedyJoinPlanOptimizerImpl extends JoinPlanOptimizerBase
+{
+    protected final FederationAccessManager fedAccessMgr;
+
+    public CardinalityBasedGreedyJoinPlanOptimizerImpl(final FederationAccessManager fedAccessMgr ) {
+        this.fedAccessMgr = fedAccessMgr;
+    }
+
+    @Override
+	public EnumerationAlgorithm initializeEnumerationAlgorithm( final List<PhysicalPlan> subplans ) {
+		return new GreedyConstructionAlgorithm(subplans);
+	}
+
+	protected class GreedyConstructionAlgorithm implements EnumerationAlgorithm {
+        protected final List<PhysicalPlan> subplans;
+
+        public GreedyConstructionAlgorithm( final List<PhysicalPlan> subplans ) {
+            this.subplans = subplans;
+        }
+
+        public PhysicalPlan getResultingPlan() throws PhysicalOptimizationException {
```

Comment:
Additionally, it would be very helpful for the future if you add a few comments into the code of this function to give our future selves or other readers of this code an idea of what is going on here. I mean, you don't need to comment every single line of code but each of the relevant steps within this functions should have a comment. As an example, you can take a look at the `apply` function of the `UnionPullUp` class.

For instance, for the first four lines below, you can add a comment such as: "Create a list of RequestMemberPair objects for all request operators inside the given list of subplans, where the list of these objects is ordered in the order in which the request operators can be found by a depth-first traversal of the subplans."